sdk/java/pulumi/src/main/java/com/pulumi/test/internal/PulumiTestInternal.java

```diff
@@ -1,34 +1,139 @@
+import com.pulumi.deployment.internal.Engine;
+import com.pulumi.deployment.internal.EngineLogger;
+import com.pulumi.deployment.internal.Monitor;
+import com.pulumi.deployment.internal.ReadOrRegisterResource;
+import com.pulumi.deployment.internal.RegisterResourceOutputs;
 import com.pulumi.deployment.internal.Runner;
 import com.pulumi.internal.PulumiInternal;
+import com.pulumi.test.EmptyMocks;
+import com.pulumi.test.Mocks;
 import com.pulumi.test.PulumiTest;
+import com.pulumi.test.TestOptions;
 import com.pulumi.test.TestResult;
 
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 import static java.util.Objects.requireNonNull;
 
+/**
+ * Provides an internal test Pulumi entrypoint and exposes various internals for the testing purposes.
+ */
+@InternalUse
 public class PulumiTestInternal extends PulumiInternal implements PulumiTest {
 
+    private final TestOptions options;
+
+    private final Log log;
     private final MockEngine engine;
     private final MockMonitor monitor;
 
+    private final ReadOrRegisterResource readOrRegisterResource;
+    private final RegisterResourceOutputs registerResourceOutputs;
+
+    @InternalUse
     public PulumiTestInternal(
+            TestOptions options,
             Runner runner,
             MockEngine engine,
             MockMonitor monitor,
+            Log log,
+            ReadOrRegisterResource readOrRegisterResource,
+            RegisterResourceOutputs registerResourceOutputs,
             ContextInternal stackContext
     ) {
         super(runner, stackContext);
+        this.options = requireNonNull(options);
+        this.log = requireNonNull(log);
         this.engine = requireNonNull(engine);
         this.monitor = requireNonNull(monitor);
+        this.readOrRegisterResource = requireNonNull(readOrRegisterResource);
+        this.registerResourceOutputs = requireNonNull(registerResourceOutputs);
+    }
+
+    /**
+     * @return the {@link TestOptions} for this test
+     */
+    public TestOptions options() {
+        return this.options;
+    }
+
+    /**
+     * @return the {@link Runner} used by the test
+     */
+    @InternalUse
+    public Runner runner() {
+        return this.runner;
+    }
+
+    /**
+     * @return the {@link Engine} used by the test
+     */
+    @InternalUse
+    public MockEngine engine() {
+        return this.engine;
+    }
+
+    /**
+     * @return return the {@link Monitor} used by the test
+     */
+    @InternalUse
+    public MockMonitor monitor() {
+        return this.monitor;
+    }
+
+    /**
+     * @return return the {@link ReadOrRegisterResource} used by the test
+     */
+    public ReadOrRegisterResource readOrRegisterResource() {
```

Comment:
used by StackTest at the moment, but I can see how this method of exposing targeted internal interfaces to the internal tests can give us scalability and freedom to refactor in the future if needed